File: contents/bogo_sort/code/go/bogo_sort.go
// Submitted by Christopher Milan (christopherm99)

package main

import (
    "fmt"
    "math/rand"
    "time"
)

func shuffle(a *[]int) {
    for i := len(*a) - 1; i > 0; i-- {
        j := rand.Intn(i + 1)
        (*a)[i], (*a)[j] = (*a)[j], (*a)[i]
    }
}

func isSorted(a []int) bool {
    for i := 0; i < len(a)-1; i++ {
        if a[i+1] < a[i] {
            return false
        }
    }
    return true
}

func bogoSort(a *[]int) {
    for !isSorted(*a) {
        shuffle(a)
    }
}

func main() {
    rand.Seed(time.Now().UnixNano())
    a := []int{1, 3, 4, 2}
    bogoSort(&a)
    fmt.Println(a)
}

File: contents/bogo_sort/code/rust/bogosort.rs
// Submitted by jess 3jane

extern crate rand;

use rand::{thread_rng, Rng};

fn is_sorted(arr: &[i32]) -> bool {
    for i in 1..arr.len() {
        if arr[i - 1] > arr[i] {
            return false;
        }
    }
    true
}

fn bogo_sort(arr: &mut [i32]) {
    while !is_sorted(arr) {
        thread_rng().shuffle(arr);
    }
}

fn main() {
    let mut v = vec![1, 2, 3, 4, 5];
    thread_rng().shuffle(&mut v);
    println!("Original array: {:?}", v);
    bogo_sort(&mut v);
    println!("Sorted array: {:?}", v);
}
